notify/rocketchat/rocketchat.go

```diff
@@ -0,0 +1,272 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rocketchat
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	commoncfg "github.com/prometheus/common/config"
+
+	"github.com/prometheus/alertmanager/config"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/alertmanager/types"
+)
+
+const maxTitleLenRunes = 1024
+
+type Notifier struct {
+	conf    *config.RocketchatConfig
+	tmpl    *template.Template
+	logger  log.Logger
+	client  *http.Client
+	retrier *notify.Retrier
+	token   string
+	tokenID string
+
+	postJSONFunc func(ctx context.Context, client *http.Client, url string, body io.Reader) (*http.Response, error)
+}
+
+// PostMessage Payload for postmessage rest API
+//
+// https://rocket.chat/docs/developer-guides/rest-api/chat/postmessage/
+type Attachment struct {
+	Title     string                              `json:"title,omitempty"`
+	TitleLink string                              `json:"title_link,omitempty"`
+	Text      string                              `json:"text,omitempty"`
+	ImageURL  string                              `json:"image_url,omitempty"`
+	ThumbURL  string                              `json:"thumb_url,omitempty"`
+	Color     string                              `json:"color,omitempty"`
+	Fields    []config.RocketchatAttachmentField  `json:"fields,omitempty"`
+	Actions   []config.RocketchatAttachmentAction `json:"actions,omitempty"`
+}
+
+// PostMessage Payload for postmessage rest API
+//
+// https://rocket.chat/docs/developer-guides/rest-api/chat/postmessage/
+type PostMessage struct {
+	Channel     string                              `json:"channel,omitempty"`
+	Text        string                              `json:"text,omitempty"`
+	ParseUrls   bool                                `json:"parseUrls,omitempty"`
+	Alias       string                              `json:"alias,omitempty"`
+	Emoji       string                              `json:"emoji,omitempty"`
+	Avatar      string                              `json:"avatar,omitempty"`
+	Attachments []Attachment                        `json:"attachments,omitempty"`
+	Actions     []config.RocketchatAttachmentAction `json:"actions,omitempty"`
+}
+
+type rocketchatRoundTripper struct {
+	wrapped http.RoundTripper
+	token   string
+	tokenID string
+}
+
+func (rcrt rocketchatRoundTripper) RoundTrip(req *http.Request) (res *http.Response, e error) {
```

Comment:
Minor change, it just avoids a couple of pointer copies:

```suggestion
func (t *rocketchatRoundTripper) RoundTrip(req *http.Request) (res *http.Response, e error) {
```